### DISCUSSION TITLE
Intercepting an uploaded file

### DISCUSSION THREAD

**Discussion by vkuznecovas** [pocketbase/pocketbase#815] [Q&A] [answered]
I'm using pocketbase as a framework, wondering if there's an easier way to gain access to a file that's been uploaded as part of a record creation.

I'd like to access the file after it was uploaded and modify the record that's being inserted into the database based on the file contents as well as return that as part of the original response.

I see that at the time of `OnRecordBeforeCreateRequest` event the file is not created yet. It's present when the `OnRecordAfterCreateRequest` gets triggered, but in my use case this is too late since the API request is long gone.

Short of reimplementing the record creation inside of `OnRecordBeforeCreateRequest` for this specific collection is there a way to achieve that?

Many thanks!

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not sure that I completely understand your use case, but if you want to access the uploaded file in the before hook, you can do that by reading it from the multipart request data. For example (_I haven't tested it_):

```go
app.OnRecordBeforeCreateRequest().Add(func(e *core.RecordCreateEvent) error {
	fh := e.HttpContext.Request().MultipartForm.File["fileField"][0]

	file, err := fh.Open()
	if err != nil {
		return err
	}
	defer file.Close()

	buf := bytes.NewBuffer(nil)
	if _, err := io.Copy(buf, file); err != nil {
		return err
	}

	fileContent := buf.Bytes()

	// update e.Record based on the fileContent ...

	return nil
})
```

**Reply by vkuznecovas** [reply]
Thanks, exactly what I was looking for!

**Reply by ganigeorgiev** [reply]
@vkuznecovas For a single file you can also use the shorter `f, fh, err := e.HttpContext.Request().FormFile("fileField")`

**Comment by dschissler** [comment]
@ganigeorgiev Is there any way that you could expose the incoming files to the model event hooks?  Since version 0.10 it's been possible to upload locally.  So that just shows that a record event is needed to be able to make it work in all environments.  That isn't what I'm trying to do now, but that just exposes a weakness.

What I'm trying to do now is to be able to run bluemonday HTML validator/cleanup on a html fragment before it's store/uploaded.  I see that I could intercept the file here, but I'd prefer to do it in the `OnModelBeforeCreate` and `OnModelBeforeUpdate` events.  I'd need to have access to the file data and to be able to manipulate it / save it again.

**Reply by ganigeorgiev** [reply]
_I guess this is related to your other question in https://github.com/pocketbase/pocketbase/discussions/1405, but unfortunately I don't fully understand what you are trying to do._

> Is there any way that you could expose the incoming files to the model event hooks?

No, the request data is available only in the `*Request` hooks. The `OnModel*` hooks respond to db events and they can be triggered outside of the HTTP request context.

**Reply by dschissler** [reply]
What about uploading locally via the new revamped system?  Is there no way to catch the file in a hook?  That seems to be incorrect.

**Reply by ganigeorgiev** [reply]
@dschissler Could you give an example what you are trying to do (pseudo-code is also fine)?

The latest local fileupload changes were related to the `forms.RecordUpsert` struct, allowing you to pass a local file (or any other file resolver struct that implements the `FileReader` interface) to the form. But there are no hooks here, if that's what you are asking.
